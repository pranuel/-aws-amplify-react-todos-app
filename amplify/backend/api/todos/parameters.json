{
    "AppSyncApiName": "todos",
    "DynamoDBBillingMode": "PAY_PER_REQUEST",
    "AuthCognitoUserPoolId": {
        "Fn::GetAtt": [
            "authtodosapp2a2fdf63",
            "Outputs.UserPoolId"
        ]
    }
}