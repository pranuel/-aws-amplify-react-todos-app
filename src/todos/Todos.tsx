import { observer } from "mobx-react";
import React from "react";
import { LoadingIndicator } from "../components/loading-indicator/loading-indicator";
import { TodoListItem } from "../components/todo-list-item";
import { TodosFilter } from "../components/todos-filter";
import { isEnterKey } from "../helpers/keyboard.helper";
import { Todo } from "./todo.model";
import { TodoStoreContract } from "./todo.store.contract";

@observer
export class Todos extends React.Component<
  { todoStore: TodoStoreContract },
  {}
> {
  public componentDidMount() {
    this.props.todoStore.initializeStore();
    this.props.todoStore.refreshTodosList();
  }

  public componentWillUnmount() {
    this.props.todoStore.dispose();
  }

  public render() {
    const {
      todos,
      newTodoDescription,
      isFetchingData: isLoading,
      itemsLeft,
      areAllTodosDone,
      currentViewMode,
      showAllTodos,
      showActiveTodos,
      showCompletedTodos,
      clearCompletedTodos,
      toggleAllTodos,
    } = this.props.todoStore;
    return (
      <div>
        <section className="todoapp">
          <header className="header">
            <h1>todos</h1>
            {isLoading ? <LoadingIndicator /> : ""}
            <input
              className="new-todo"
              placeholder="What needs to be done?"
              value={newTodoDescription}
              onChange={this.onChangeNewTodoInput}
              onKeyDown={this.onNewTodoKeyDown}
            />
          </header>
          <section className="main">
            <input
              id="toggle-all"
              className="toggle-all"
              type="checkbox"
              checked={areAllTodosDone}
              onChange={toggleAllTodos}
            />
            <label htmlFor="toggle-all" />
            <ul className="todo-list">{todos.map(this.getTodoListItem)}</ul>
          </section>
          <footer className="footer">
            <span className="todo-count">
              <strong>{itemsLeft}</strong>
              <span />
              <span>items</span>
              <span> left</span>
            </span>
            <TodosFilter
              currentViewMode={currentViewMode}
              onShowActiveTodosClicked={showActiveTodos}
              onShowAllTodosClicked={showAllTodos}
              onShowCompletedTodosClicked={showCompletedTodos}
            />
            <button className="clear-completed" onClick={clearCompletedTodos}>
              Clear completed
            </button>
          </footer>
        </section>
        <footer className="info">
          <p>Double-click to edit a todo</p>
          <p>
            Created by <a href="https://github.com/pranuel">Pranuel</a>
          </p>
          <p>
            Part of <a href="http://todomvc.com">TodoMVC</a>
          </p>
        </footer>
      </div>
    );
  }

  private getTodoListItem = (todo: Todo) => {
    const {
      editTodo,
      deleteTodo,
      setEditTodo,
      changeEditingTodoDescription,
      saveTodo,
      updateTodoIsDone,
    } = this.props.todoStore;
    return (
      <TodoListItem
        key={todo.id}
        todo={todo}
        editTodo={editTodo}
        onDeleteTodo={deleteTodo}
        onEditTodo={setEditTodo}
        onEditingTodoDescriptionChanged={changeEditingTodoDescription}
        onSaveTodo={saveTodo}
        onTodoIsDoneToggled={updateTodoIsDone}
      />
    );
  }

  private onChangeNewTodoInput = (
    event: React.ChangeEvent<HTMLInputElement>,
  ) => {
    this.props.todoStore.setNewTodoDescription(event.target.value);
  }

  private onNewTodoKeyDown(event: React.KeyboardEvent<HTMLInputElement>) {
    if (!isEnterKey(event.key)) {
      return;
    }
    this.props.todoStore.createTodo(this.props.todoStore.newTodoDescription);
  }
}
