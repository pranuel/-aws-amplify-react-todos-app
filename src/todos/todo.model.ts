import { observable } from "mobx";

export class Todo {
  public id = "";
  @observable public isDone = false;
  @observable public description = "";
}
