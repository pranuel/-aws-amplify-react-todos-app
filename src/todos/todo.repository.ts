import { GraphQLResult } from "@aws-amplify/api/lib/types";
import { API, graphqlOperation } from "aws-amplify";
import {
  CreateTodoMutationVariables,
  DeleteTodoInput,
  ListTodosQuery,
  ModelTodoFilterInput,
  UpdateTodoInput,
} from "../API";
import * as Mutations from "../graphql/mutations";
import { listTodos } from "../graphql/queries";
import { Todo } from "./todo.model";

/*
# API
*/

export class TodoRepository {
  public async getAllTodos() {
    return this.getTodos();
  }

  public async getCompletedTodos() {
    return this.getTodos({
      isDone: { eq: true },
    });
  }

  public async getActiveTodos() {
    return this.getTodos({
      isDone: { eq: false },
    });
  }

  public async createTodo(description: string, isDone: boolean) {
    const createTodoMutationVariables: CreateTodoMutationVariables = {
      input: {
        description,
        isDone,
      },
    };
    await API.graphql(
      graphqlOperation(Mutations.createTodo, createTodoMutationVariables),
    );
  }

  public async updateTodo(todo: Todo) {
    const input: UpdateTodoInput = todo;
    await API.graphql(graphqlOperation(Mutations.updateTodo, { input }));
  }

  public async deleteTodo(todo: Todo) {
    const input: DeleteTodoInput = {
      id: todo.id,
    };
    await API.graphql(graphqlOperation(Mutations.deleteTodo, { input }));
  }

  private async getTodos(filter?: ModelTodoFilterInput): Promise<Todo[]> {
    const allTodosResult = await API.graphql(
      graphqlOperation(listTodos, { filter }),
    );
    if (this.isGraphQLResult(allTodosResult)) {
      const listTodosQuery = allTodosResult.data as ListTodosQuery;
      if (!!listTodosQuery.listTodos) {
        const todos = listTodosQuery.listTodos.items as Todo[];
        return todos;
      }
    }
    return [];
  }

  private isGraphQLResult(result: {}): result is GraphQLResult {
    // this crazy stuff was taken from here: https://www.typescriptlang.org/docs/handbook/advanced-types.html
    return (result as GraphQLResult).data !== undefined;
  }
}
