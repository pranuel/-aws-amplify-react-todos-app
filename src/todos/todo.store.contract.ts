import { Todo } from "./todo.model";
import { ViewModes } from "./viewmodes.model";

export interface TodoStoreContract {
  readonly areAllTodosDone: boolean;
  readonly itemsLeft: number;
  readonly completedTodos: Todo[];

  todos: Todo[];
  newTodoDescription: string;
  isFetchingData: boolean;
  editTodo?: Todo;
  currentViewMode: ViewModes;
  error?: string;

  dispose: () => void;
  setEditTodo: (editTodo: Todo) => void;
  setNewTodoDescription: (description: string) => void;
  refreshTodosList: () => Promise<void>;
  refreshActiveTodos: () => Promise<void>;
  refreshCompletedTodos: () => Promise<void>;
  createTodo: (description: string) => Promise<void>;
  deleteTodo: (todo: Todo) => Promise<void>;
  updateTodoIsDone: (todo: Todo, isDone: boolean) => Promise<void>;
  saveTodo: (todo: Todo) => Promise<void>;
  changeEditingTodoDescription: (editTodo?: Todo, description?: string) => void;
  toggleAllTodos: () => Promise<void>;
  showAllTodos: () => void;
  showActiveTodos: () => void;
  showCompletedTodos: () => void;
  clearCompletedTodos: () => Promise<void>;
  initializeStore: () => void;
}
