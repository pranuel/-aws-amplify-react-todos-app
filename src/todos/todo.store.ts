import { action, autorun, computed, IReactionDisposer, observable } from "mobx";
import { TodoFilter } from "./todo-filter.model";
import { Todo } from "./todo.model";
import { TodoRepository } from "./todo.repository";
import { TodoStoreContract } from "./todo.store.contract";

export class TodoStore implements TodoStoreContract {
  @computed get areAllTodosDone() {
    return this.todos.every((todo) => todo.isDone);
  }

  @computed get itemsLeft() {
    return this.todos.filter((todo) => !todo.isDone).length;
  }

  @computed get completedTodos() {
    return this.todos.filter((todo) => todo.isDone);
  }

  @observable public todos: Todo[];
  @observable public newTodoDescription: string;
  @observable public isFetchingData: boolean;
  @observable public editTodo?: Todo;
  @observable public todoFilter: TodoFilter;
  @observable public error?: string;
  private readonly autoRunDisposers: IReactionDisposer[];

  constructor(private readonly todoRepository: TodoRepository) {
    this.todos = [];
    this.newTodoDescription = "";
    this.isFetchingData = false;
    this.todoFilter = TodoFilter.All;
    this.autoRunDisposers = [];
  }

  public dispose() {
    this.autoRunDisposers.forEach((disposer) => {
      disposer();
    });
  }

  @action
  public setEditTodo = (editTodo: Todo) => {
    this.editTodo = editTodo;
  }

  @action
  public setNewTodoDescription = (description: string) => {
    this.newTodoDescription = description;
  }

  @action
  public createTodo = async (description: string) => {
    this.resetNewTodoProperties();
    this.isFetchingData = true;
    await this.todoRepository.createTodo(description, false);
    await this.refreshTodosList();
    this.isFetchingData = false;
  }

  @action
  public deleteTodo = async (todo: Todo) => {
    this.isFetchingData = true;
    await this.todoRepository.deleteTodo(todo);
    await this.refreshTodosList();
    this.isFetchingData = false;
  }

  @action
  public updateTodoIsDone = async (todo: Todo, isDone: boolean) => {
    todo.isDone = isDone;
    this.isFetchingData = true;
    await this.todoRepository.updateTodo(todo);
    this.isFetchingData = false;
  }

  @action
  public saveTodo = async (todo: Todo) => {
    this.isFetchingData = true;
    await this.todoRepository.updateTodo(todo);
    this.isFetchingData = false;
    this.editTodo = undefined;
  }

  @action
  public changeEditingTodoDescription = (
    editTodo?: Todo,
    description?: string,
  ) => {
    if (!editTodo) {
      return;
    }
    editTodo.description = description || "";
  }

  @action
  public toggleAllTodos = async () => {
    this.isFetchingData = true;
    const areAllTodosDone = this.areAllTodosDone;
    const updateTodoPromises = this.todos.map((todo) => {
      todo.isDone = !areAllTodosDone;
      return this.todoRepository.updateTodo(todo);
    });
    await Promise.all(updateTodoPromises);
    this.isFetchingData = false;
  }

  @action
  public showAllTodos = () => {
    this.todoFilter = TodoFilter.All;
  }

  @action
  public showActiveTodos = () => {
    this.todoFilter = TodoFilter.Active;
  }

  @action
  public showCompletedTodos = () => {
    this.todoFilter = TodoFilter.Completed;
  }

  @action
  public clearCompletedTodos = async () => {
    const deleteCompletedTodoPromises = this.completedTodos.map(
      this.deleteTodo.bind(this),
    );
    this.isFetchingData = true;
    await Promise.all(deleteCompletedTodoPromises);
    await this.refreshTodosList();
    this.isFetchingData = false;
  }

  public initializeStore() {
    const reactOnViewModeDisposer = autorun(async () => {
      await this.refreshTodosList();
    });
    this.autoRunDisposers.push(reactOnViewModeDisposer);
  }

  public async refreshTodosList() {
    switch (this.todoFilter) {
      case TodoFilter.All:
        await this.refreshAllTodos();
        break;
      case TodoFilter.Active:
        await this.refreshActiveTodos();
        break;
      case TodoFilter.Completed:
        await this.refreshCompletedTodos();
        break;
    }
  }

  @action
  private refreshAllTodos = async () => {
    this.isFetchingData = true;
    try {
      this.todos = await this.todoRepository.getAllTodos();
    } catch (error) {
      this.error = "Error during getAllTodos(): " + error;
    } finally {
      this.isFetchingData = false;
    }
  }

  @action
  private refreshActiveTodos = async () => {
    this.isFetchingData = true;
    try {
      this.todos = await this.todoRepository.getActiveTodos();
    } catch (error) {
      this.error = "Error during fetchActiveTodos(): " + error;
    } finally {
      this.isFetchingData = false;
    }
  }

  @action
  private refreshCompletedTodos = async () => {
    this.isFetchingData = true;
    try {
      this.todos = await this.todoRepository.getCompletedTodos();
    } catch (error) {
      this.error = "Error during fetchCompletedTodos(): " + error;
    } finally {
      this.isFetchingData = false;
    }
  }

  @action
  private resetNewTodoProperties = () => {
    this.setNewTodoDescription("");
  }
}
