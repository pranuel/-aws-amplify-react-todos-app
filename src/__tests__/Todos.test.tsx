import React from "react";
import renderer from "react-test-renderer";
import { TodoStoreContract } from "../todos/todo.store.contract";
import { Todos } from "../todos/Todos";
import { ViewModes } from "../todos/viewmodes.model";

describe("Todos", () => {
  it("should render correctly", () => {
    const mockedTodoStore: TodoStoreContract = {
      // public props
      areAllTodosDone: false,
      completedTodos: [],
      currentViewMode: ViewModes.All,
      editTodo: undefined,
      error: undefined,
      isLoading: false,
      itemsLeft: 2,
      newTodoDescription: "foo",
      todos: [
        {
          description: "bar description",
          id: "1",
          isDone: false,
        },
        {
          description: "foo description",
          id: "2",
          isDone: false,
        },
      ],
      // public methods
      initializeStore: jest.fn(),
      refreshTodosList: jest.fn(),
      changeEditingTodoDescription: jest.fn(),
      clearCompletedTodos: jest.fn(),
      createTodo: jest.fn(),
      deleteTodo: jest.fn(),
      dispose: jest.fn(),
      refreshActiveTodos: jest.fn(),
      refreshCompletedTodos: jest.fn(),
      saveTodo: jest.fn(),
      setEditTodo: jest.fn(),
      setNewTodoDescription: jest.fn(),
      showActiveTodos: jest.fn(),
      showAllTodos: jest.fn(),
      showCompletedTodos: jest.fn(),
      toggleAllTodos: jest.fn(),
      updateTodoIsDone: jest.fn(),
    };
    const tree = renderer
      .create(<Todos todoStore={mockedTodoStore} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
