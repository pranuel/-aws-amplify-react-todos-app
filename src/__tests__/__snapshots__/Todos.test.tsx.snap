// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todos should render correctly 1`] = `
<div>
  <section
    className="todoapp"
  >
    <header
      className="header"
    >
      <h1>
        todos
      </h1>
      
      <input
        className="new-todo"
        onChange={[Function]}
        onKeyDown={[Function]}
        placeholder="What needs to be done?"
        value="foo"
      />
    </header>
    <section
      className="main"
    >
      <input
        checked={false}
        className="toggle-all"
        id="toggle-all"
        onChange={[MockFunction]}
        type="checkbox"
      />
      <label
        htmlFor="toggle-all"
      />
      <ul
        className="todo-list"
      >
        <li
          className=""
          onDoubleClick={[Function]}
        >
          <div
            className="view"
          >
            <input
              checked={false}
              className="toggle"
              data-testid="is-done-toggle"
              onChange={[Function]}
              type="checkbox"
            />
            <label>
              bar description
            </label>
            <button
              className="destroy"
              data-testid="delete-button"
              onClick={[Function]}
            />
          </div>
          <input
            className="edit"
            onBlur={[Function]}
            onChange={[Function]}
            onKeyDown={[Function]}
            value=""
          />
        </li>
        <li
          className=""
          onDoubleClick={[Function]}
        >
          <div
            className="view"
          >
            <input
              checked={false}
              className="toggle"
              data-testid="is-done-toggle"
              onChange={[Function]}
              type="checkbox"
            />
            <label>
              foo description
            </label>
            <button
              className="destroy"
              data-testid="delete-button"
              onClick={[Function]}
            />
          </div>
          <input
            className="edit"
            onBlur={[Function]}
            onChange={[Function]}
            onKeyDown={[Function]}
            value=""
          />
        </li>
      </ul>
    </section>
    <footer
      className="footer"
    >
      <span
        className="todo-count"
      >
        <strong>
          2
        </strong>
        <span />
        <span>
          items
        </span>
        <span>
           left
        </span>
      </span>
      <ul
        className="filters"
      >
        <li>
          <a
            className="selected"
            href="#/"
            onClick={[MockFunction]}
          >
            All
          </a>
        </li>
        <span />
        <li>
          <a
            href="#/"
            onClick={[MockFunction]}
          >
            Active
          </a>
        </li>
        <span />
        <li>
          <a
            href="#/"
            onClick={[MockFunction]}
          >
            Completed
          </a>
        </li>
      </ul>
      <button
        className="clear-completed"
        onClick={[MockFunction]}
      >
        Clear completed
      </button>
    </footer>
  </section>
  <footer
    className="info"
  >
    <p>
      Double-click to edit a todo
    </p>
    <p>
      Created by 
      <a
        href="https://github.com/pranuel"
      >
        Pranuel
      </a>
    </p>
    <p>
      Part of 
      <a
        href="http://todomvc.com"
      >
        TodoMVC
      </a>
    </p>
  </footer>
</div>
`;
